### PR TITLE
Allow loading triangular meshes from Gmsh with some cells inverted

### PR DESCRIPTION
**Pull request by zjiaqi2018**
Related to https://github.com/dealii/dealii/issues/13346

Also tested a few tetrahedron meshes generated by Gmsh, they don't seem to have this problem,
so I didn't add those 3D tests. 
@tjhei

### CODE REVIEW

File: source/grid/grid_tools.cc [hunk 1]
```diff
@@ -872,34 +872,60 @@ namespace GridTools
...
-            // TODO: this only works for quads and hexes
-            if (dim == 2)
+            const ArrayView<const unsigned int> vertices(cell.vertices);
+            if (GridTools::cell_measure(all_vertices, vertices) < 0)
               {
-                // flip the cell across the y = x line in 2D
-                std::swap(cell.vertices[1], cell.vertices[2]);
+                ++n_negative_cells;
+
+                // TODO: this only works for quads and hexes
```
Comment (tjhei) [inline source/grid/grid_tools.cc:883]: this comment can be removed, right?

File: source/grid/grid_tools.cc [hunk 2]
```diff
@@ -872,34 +872,60 @@ namespace GridTools
...
-                std::swap(cell.vertices[0], cell.vertices[2]);
-                std::swap(cell.vertices[1], cell.vertices[3]);
-                std::swap(cell.vertices[4], cell.vertices[6]);
-                std::swap(cell.vertices[5], cell.vertices[7]);
+                ++n_negative_cells;
+                if (dim == 2)
+                  {
+                    // Triangular mesh, swap any two vertices
+                    std::swap(cell.vertices[1], cell.vertices[2]);
+                    AssertThrow(GridTools::cell_measure(all_vertices,
```
Comment (tjhei) [inline source/grid/grid_tools.cc:916]: you can move this check outside the if (and remove the duplicate above)

File: source/grid/grid_tools.cc [hunk 3]
```diff
@@ -872,34 +872,60 @@ namespace GridTools
...
+                                                        vertices) > 0,
+                                ExcInternalError());
+                  }
+                else
+                  {
+                    Assert(false, ExcNotImplemented());
```
Comment (tjhei) [inline source/grid/grid_tools.cc:922]: This should be an AssertThrow.

File: source/grid/grid_tools.cc [hunk 4]
```diff
@@ -872,34 +872,60 @@ namespace GridTools
...
-
-            // Check whether the resulting cell is now ok.
-            // If not, then the grid is seriously broken and
-            // we just give up.
-            AssertThrow(GridTools::cell_measure(all_vertices, vertices) > 0,
-                        ExcInternalError());
+          }
+        else
+          {
+            Assert(false, ExcNotImplemented());
```
Comment (tjhei) [inline source/grid/grid_tools.cc:928]: This should also be an AssertThrow. The reason is that this is triggered by user-data.

File: tests/grid/grid_in_gmsh_03.output [hunk 5]
```diff
@@ -0,0 +1,7 @@
+
+DEAL::Mesh info:
+DEAL:: dimension: 2
+DEAL:: no. of cells: 984
```
Comment (tjhei) [inline tests/grid/grid_in_gmsh_03.output:4]: Does the mesh need to be this big? I am sure you can reproduce this with a handful of cells, right?

File: source/grid/grid_tools_nontemplates.cc [hunk 6]
```diff
@@ -57,8 +57,8 @@ namespace GridTools
                              all_vertices[vertex_indices[1]](1),
                              all_vertices[vertex_indices[2]](1)};
 
-        return 0.5 * std::abs((x[0] - x[2]) * (y[1] - y[0]) -
-                              (x[0] - x[1]) * (y[2] - y[0]));
+        return 0.5 *
```
Comment (tjhei) [inline source/grid/grid_tools_nontemplates.cc:60]: I am not sure why abs() was used here before. @peterrum , do you know why?

**Review by tjhei — APPROVED** [review]
Awesome. Can you add a changelog entry please?

**Comment by tjhei** [comment]
Did you run the simplex tests?

**Comment by zjiaqi2018** [comment]
> Did you run the simplex tests?

Yes, all passed.

**Comment by tjhei** [comment]
Let us know once you have addressed all comments.

File: source/grid/grid_tools_nontemplates.cc (hunk 6 above)
Comment (peterrum) [reply source/grid/grid_tools_nontemplates.cc:60]: I guess the person who wrote this (probably me) did not think of negative determinants^^

**Comment by zjiaqi2018** [comment]
> Let us know once you have addressed all comments.

I think I have addressed all the comments.

File: source/grid/grid_tools.cc [hunk 7]
```diff
@@ -864,6 +864,17 @@ namespace GridTools
     const std::vector<Point<spacedim>> &all_vertices,
     std::vector<CellData<dim>> &        cells)
   {
+    // This only works for quads and hexes,
+    // and triangules. From the examples we tested,
```
Comment (bangerth) [inline source/grid/grid_tools.cc:868]:
```suggestion
    // and triangles. From the examples we tested,
```

File: source/grid/grid_tools.cc [hunk 8]
```diff
@@ -864,6 +864,17 @@ namespace GridTools
...
+    // tetrahedron meshes loaded from Gmsh don't have
+    // this problem, so we skip them for now.
+
+    // Since we use std::abs in computing the measures
+    // of tetrahdrons, the measures are always positive.
```
Comment (bangerth) [inline source/grid/grid_tools.cc:873]:
```suggestion
    // of tetrahedra, the measures are always positive.
```

File: source/grid/grid_tools.cc [hunk 9]
```diff
@@ -864,6 +864,17 @@ namespace GridTools
...
+    // But if the orientation is wrong, we may get negative
+    // determinants of the Jacobians in MappingFE, which
+    // will triger the assert there.
```
Comment (bangerth) [inline source/grid/grid_tools.cc:876]: It's not clear to me what this comment refers to, in particular the "This only works..." at the very beginning. Are you talking about the algorithms used here, the whole function, or something else?

You're also mentioning a call to `abs()`, but there is none in the following several lines of code.

File: source/grid/grid_tools.cc [hunk 10]
```diff
@@ -872,35 +883,55 @@ namespace GridTools
...
         const ArrayView<const unsigned int> vertices(cell.vertices);
-        if (GridTools::cell_measure(all_vertices, vertices) < 0)
+        if (cell.vertices.size() ==
+            ReferenceCells::get_hypercube<dim>().n_vertices())
           {
-            ++n_negative_cells;
-
-            // TODO: this only works for quads and hexes
-            if (dim == 2)
+            if (GridTools::cell_measure(all_vertices, vertices) < 0)
```
Comment (bangerth) [inline source/grid/grid_tools.cc:890]: I would pull the call to `cell_measure()` out front, and only then when this call return true have the `if` statements about the reference cell.

You can write the call to figure out which reference cell you have in an easier to read way if you use `if (ReferenceCell::n_vertices_to_type(dim, vertices.size()).is_hyper_cube())` and similarly for `...is_simplex()`.

**Review by bangerth — CHANGES_REQUESTED** [review]

File: source/grid/grid_tools.cc (hunk 9 above)
Comment (bangerth) [reply source/grid/grid_tools.cc:876]: About "This...": One of the things my wife always complains about when she looks at my writing is the use of "this" on its own. One typically uses it to reference something in the previous sentence, but it's not always clear what that is. She always requests that I write "This algorithm...", "this observation...", or whatever else it is that I refer to.

Here, you have no previous sentence at all, so it is doubly unclear what is meant :-)

**Review by kronbichler — APPROVED** [review]

**Pull request merged**